File: PE-common/src/main/java/nl/adaptivity/process/processModel/JoinBase.java
package nl.adaptivity.process.processModel;

import nl.adaptivity.process.util.Identifiable;
import nl.adaptivity.process.util.Identifier;
import nl.adaptivity.util.xml.XmlUtil;
import nl.adaptivity.xml.XmlException;
import nl.adaptivity.xml.XmlReader;
import nl.adaptivity.xml.XmlWriter;
import org.jetbrains.annotations.NotNull;

import javax.xml.namespace.QName;

import java.util.Collection;


/**
 * Created by pdvrieze on 26/11/15.
 */
public class JoinBase<T extends ProcessNode<T, M>, M extends ProcessModelBase<T,M>> extends JoinSplitBase<T, M> implements Join<T, M> {

  public static final String IDBASE = "join";

  public JoinBase(final M ownerModel, final Collection<? extends Identifiable> predecessors, final int max, final int min) {super(ownerModel, predecessors, max, min);}

  public JoinBase(final M ownerModel) {super(ownerModel);}

  public JoinBase(final Join<?, ?> orig) {
    super(orig);
  }

  @Override
  public final String getIdBase() {
    return IDBASE;
  }

  @Override
  public void serialize(@NotNull final XmlWriter out) throws XmlException {
    XmlUtil.writeStartElement(out, ELEMENTNAME);
    serializeAttributes(out);
    serializeChildren(out);
    XmlUtil.writeEndElement(out, ELEMENTNAME);
  }

  protected void serializeChildren(@NotNull final XmlWriter out) throws XmlException {
    super.serializeChildren(out);
    for(final Identifiable pred: getPredecessors()) {
      XmlUtil.writeStartElement(out, PREDELEMNAME);
      out.text(pred.getId());
      XmlUtil.writeEndElement(out, PREDELEMNAME);
    }
  }

  @Override
  public boolean deserializeChild(@NotNull final XmlReader in) throws XmlException {
    if (XmlUtil.isElement(in, PREDELEMNAME)) {
      final String id = XmlUtil.readSimpleElement(in).toString();
      addPredecessor(new Identifier(id));
      return true;
    }
    return super.deserializeChild(in);
  }

  @Override
  public final <R> R visit(@NotNull final Visitor<R> visitor) {
    return visitor.visitJoin(this);
  }

  @NotNull
  @Override
  public final QName getElementName() {
    return ELEMENTNAME;
  }

  @Override
  public final int getMaxPredecessorCount() {
    return Integer.MAX_VALUE;
  }
}

File: PE-common/src/main/kotlin/nl/adaptivity/process/processModel/JoinBase.kt
/*
 * Copyright (c) 2016.
 *
 * This file is part of ProcessManager.
 *
 * ProcessManager is free software: you can redistribute it and/or modify it under the terms of version 3 of the
 * GNU Lesser General Public License as published by the Free Software Foundation.
 *
 * ProcessManager is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even
 * the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License along with ProcessManager.  If not,
 * see <http://www.gnu.org/licenses/>.
 */

package nl.adaptivity.process.processModel

import nl.adaptivity.process.util.Identified
import nl.adaptivity.process.util.Identifier
import nl.adaptivity.xml.*
import javax.xml.namespace.QName


/**
 * Created by pdvrieze on 26/11/15.
 */
abstract class JoinBase<NodeT : ProcessNode<NodeT, ModelT>, ModelT : ProcessModel<NodeT, ModelT>?> : JoinSplitBase<NodeT, ModelT>, Join<NodeT, ModelT> {

  abstract class Builder<NodeT : ProcessNode<NodeT, ModelT>, ModelT : ProcessModel<NodeT, ModelT>?> : JoinSplitBase.Builder<NodeT,ModelT>, Join.Builder<NodeT,ModelT> {
    override val idBase:String
      get() = "join"

    override var isMultiMerge: Boolean

    constructor():this(predecessors= emptyList(), isMultiMerge = false, isMultiInstance = false)

    constructor(id: String? = null,
                predecessors: Collection<Identified> = emptyList(),
                successor: Identified? = null, label: String? = null,
                defines: Collection<IXmlDefineType> = emptyList(),
                results: Collection<IXmlResultType> = emptyList(),
                x: Double = Double.NaN,
                y: Double = Double.NaN,
                min: Int = -1,
                max: Int = -1,
                isMultiMerge: Boolean = false,
                isMultiInstance: Boolean = false) : super(id, predecessors, listOfNotNull(successor), label, defines, results, x,
                                                          y, min, max, isMultiInstance) {
      this.isMultiMerge = isMultiMerge
    }

    constructor(node: Join<*, *>) : super(node) {
      this.isMultiMerge = node.isMultiMerge
    }

    @Throws(XmlException::class)
    override fun deserializeChild(reader: XmlReader): Boolean {
      if (reader.isElement(Join.PREDELEMNAME)) {
        val id = reader.readSimpleElement().toString()
        predecessors.add(Identifier(id))
        return true
      }
      return super.deserializeChild(reader)
    }

    override val elementName: QName
      get() = Join.ELEMENTNAME

  }

  @Deprecated("")
  constructor(ownerModel: ModelT) : super(ownerModel) {
    isMultiMerge = false
  }

  constructor(builder: Join.Builder<*, *>, buildHelper: ProcessModel.BuildHelper<NodeT, ModelT>) : super(builder, buildHelper) {
    isMultiMerge = builder.isMultiMerge
  }

  override abstract fun builder(): Builder<NodeT, ModelT>

  override val idBase: String
    get() = IDBASE

  override val isMultiMerge: Boolean

  @Throws(XmlException::class)
  override fun serialize(out: XmlWriter) {
    out.smartStartTag(Join.ELEMENTNAME)
    serializeAttributes(out)
    serializeChildren(out)
    out.endTag(Join.ELEMENTNAME)
  }

  @Throws(XmlException::class)
  override fun serializeChildren(out: XmlWriter) {
    super.serializeChildren(out)
    for (pred in predecessors) {
      out.smartStartTag(Join.PREDELEMNAME)
      out.text(pred.id)
      out.endTag(Join.PREDELEMNAME)
    }
  }

  override fun <R> visit(visitor: ProcessNode.Visitor<R>): R {
    return visitor.visitJoin(this)
  }

  override val maxPredecessorCount: Int
    get() = Integer.MAX_VALUE

  companion object {

    val IDBASE = "join"
  }
}
